File: python/Command/remote_control.py

from typing import List
from package.command import Command, NoCommand
import copy


class RemoteControl:
    def __init__(self) -> None:
        self.solt_num: int = 7
        self.on_commands: List[Command]
        self.off_commands: List[Command]
        self.undo_command_stack: List[Command] = []

        self.on_commands = [NoCommand() for i in range(self.solt_num)]
        self.off_commands = [NoCommand() for i in range(self.solt_num)]

    def set_command(self, solt: int, on_command: Command, off_command: Command) -> None:
        self.on_commands[solt] = on_command
        self.off_commands[solt] = off_command

    def on_button_was_pushed(self, solt: int) -> None:
        self.on_commands[solt].execute()
        self.undo_command_stack.append(copy.copy(self.on_commands[solt]))

    def off_button_was_pushed(self, solt: int) -> None:
        self.off_commands[solt].execute()
        self.undo_command_stack.append(copy.copy(self.off_commands[solt]))

    def undo_button_was_pushed(self) -> None:
        if len(self.undo_command_stack) == 0:
            print("there is nothing undo\n")
            return
        undo_command = self.undo_command_stack.pop()
        undo_command.undo()

    def __str__(self) -> str:
        string = "\n----- Remote Control -----\n"
        for i in range(self.solt_num):
            string += f"[solt {i}] {self.on_commands[i].__class__.__name__}\t{self.off_commands[i].__class__.__name__}\n"
        return string

File: c/Command/remote_control.c
#include "remote_control.h"
#include <stdio.h>

Command no_command;

void RemoteControlNew(RemoteControl *self)
{
    NoCommandNew(&no_command);
    for (int i = 0; i < SOLT_NUM; i++)
    {
        self->on_commands[i] = &no_command;
        self->off_commands[i] = &no_command;
    }
    for (int i = 0; i < UNDO_SIZE; i++)
    {
        self->undo_command_stack[i] = &no_command;
    }
    self->undo_idx = -1;
    self->undo_len = 0;
    self->set_command = RC_set_command;
    self->on_button_was_pushed = RC_on_button_was_pushed;
    self->off_button_was_pushed = RC_off_button_was_pushed;
    self->undo_button_was_pushed = RC_undo_button_was_pushed;
    self->print = RC_print;
}
void RC_set_command(RemoteControl *self, int solt, Command *on_command, Command *off_command)
{
    self->on_commands[solt] = on_command;
    self->off_commands[solt] = off_command;
}
void RC_on_button_was_pushed(RemoteControl *self, int solt)
{
    self->on_commands[solt]->execute(self->on_commands[solt]);
    if (self->undo_len < UNDO_SIZE)
    {
        self->undo_idx++;
        self->undo_len++;
        self->undo_command_stack[self->undo_idx] = self->on_commands[solt];
    }
}
void RC_off_button_was_pushed(RemoteControl *self, int solt)
{
    self->off_commands[solt]->execute(self->off_commands[solt]);
    if (self->undo_len < UNDO_SIZE)
    {
        self->undo_idx++;
        self->undo_len++;
        self->undo_command_stack[self->undo_idx] = self->off_commands[solt];
    }
}
void RC_undo_button_was_pushed(RemoteControl *self)
{
    if (self->undo_len > 0)
    {
        Command *undo_command = self->undo_command_stack[self->undo_idx];
        undo_command->undo(undo_command);
        self->undo_idx--;
        self->undo_len--;
    }
}
void RC_print(RemoteControl *self)
{
    printf("\n----- Remote Control -----\n");
    for (int i = 0; i < SOLT_NUM; i++)
    {
        printf("[solt %d] %-25s %s\n", i, self->on_commands[i]->name, self->off_commands[i]->name);
    }
}
